File: backend/app/runtime/task.py
from dataclasses import dataclass, field
from datetime import datetime, timezone
from typing import Any
from uuid import uuid4


@dataclass
class Task:
    """
    AI-Commerce-OS 标准任务对象。
    """

    task_type: str
    payload: dict[str, Any]

    assigned_agent: str | None = None
    priority: str = "normal"

    id: str = field(
        default_factory=lambda: f"TASK-{uuid4().hex[:12].upper()}"
    )

    status: str = "pending"

    created_at: datetime = field(
        default_factory=lambda: datetime.now(timezone.utc)
    )

    started_at: datetime | None = None
    completed_at: datetime | None = None

    result: dict[str, Any] | None = None
    error: str | None = None

    # 阶段 8B：父子任务委派。普通（非委派）任务 parent_task_id 为
    # None、delegation_depth=0；root_task_id 在 __post_init__ 里
    # 统一默认回填为自身 id（不区分"根任务"和"普通任务"两套语义，
    # 任何任务都能安全地用 root_task_id 查询整条委派链）。
    # AI CEO 创建子任务时会显式传入 parent_task_id/root_task_id/
    # delegation_depth=1/created_by_agent/delegation_key，不使用
    # 这里的默认值。
    parent_task_id: str | None = None
    root_task_id: str | None = None
    delegation_depth: int = 0
    created_by_agent: str | None = None
    delegation_key: str | None = None

    # 阶段 8E：店铺业务作用域，None 表示"未绑定店铺"。
    shop_id: int | None = None
    source_deliverable_id: int | None = None

    def __post_init__(self) -> None:
        if self.root_task_id is None:
            self.root_task_id = self.id

    def mark_running(self) -> None:
        """
        将任务标记为执行中。
        """

        self.status = "running"
        self.started_at = datetime.now(timezone.utc)
        self.completed_at = None
        self.error = None

    def mark_completed(
        self,
        result: dict[str, Any],
    ) -> None:
        """
        将任务标记为已完成。
        """

        self.status = "completed"
        self.result = result
        self.completed_at = datetime.now(timezone.utc)
        self.error = None

    def mark_failed(
        self,
        error: str,
    ) -> None:
        """
        将任务标记为执行失败。
        """

        self.status = "failed"
        self.error = error
        self.completed_at = datetime.now(timezone.utc)

    def to_dict(self) -> dict[str, Any]:
        """
        输出适用于 API 的任务数据。
        """

        return {
            "id": self.id,
            "task_type": self.task_type,
            "payload": self.payload,
            "assigned_agent": self.assigned_agent,
            "priority": self.priority,
            "status": self.status,
            "created_at": self.created_at.isoformat(),
            "started_at": (
                self.started_at.isoformat()
                if self.started_at
                else None
            ),
            "completed_at": (
                self.completed_at.isoformat()
                if self.completed_at
                else None
            ),
            "result": self.result,
            "error": self.error,
            "parent_task_id": self.parent_task_id,
            "root_task_id": self.root_task_id,
            "delegation_depth": self.delegation_depth,
            "created_by_agent": self.created_by_agent,
            "shop_id": self.shop_id,
            "source_deliverable_id": self.source_deliverable_id,
        }

File: frontend/src/types/task.js
/**
 * Task 提交相关的响应结构定义（JSDoc 类型）。
 *
 * 项目当前未启用 TypeScript（无 tsconfig，全部源码为 .js/.jsx），
 * 这里用 JSDoc typedef 记录与后端 POST /api/v1/tasks/submit 完全
 * 一致的字段形状，供编辑器提示与代码阅读参考，不引入新的构建
 * 工具链。
 *
 * @typedef {"high" | "normal" | "low"} TaskPriority
 */

/**
 * POST /api/v1/tasks/submit 的请求体。
 *
 * context 本轮不向用户开放编辑，前端固定发送空对象。
 *
 * @typedef {Object} TaskSubmitRequest
 * @property {string} assigned_agent
 * @property {string} task
 * @property {Object.<string, any>} context
 * @property {TaskPriority} priority
 * @property {number | null} [shop_id] 阶段 8E：目标店铺 id，null 表示未绑定店铺
 */

/**
 * POST /api/v1/tasks/submit 的成功响应（HTTP 202）。
 *
 * status 恒为 "pending"——本接口只入队，不代表任务已经执行完成。
 *
 * @typedef {Object} TaskSubmitResponse
 * @property {string} id
 * @property {string} status
 * @property {string} assigned_agent
 * @property {string} task_type
 * @property {TaskPriority} priority
 * @property {string} created_at
 * @property {string} message
 */

/**
 * GET /api/v1/tasks/{task_id}（以及 GET /api/v1/tasks 列表项）返回的
 * 完整任务结构。展示到 UI 前必须先经过
 * taskDetailHelpers.sanitizeTaskDetail() 处理，不得把该对象整体
 * （尤其是 payload/context）直接传给界面。
 *
 * @typedef {Object} TaskDetail
 * @property {string} id
 * @property {string} task_type
 * @property {Object.<string, any>} payload
 * @property {string | null} assigned_agent
 * @property {string} priority
 * @property {string} status
 * @property {string | null} created_at
 * @property {string | null} started_at
 * @property {string | null} completed_at
 * @property {Object.<string, any> | null} result
 * @property {string | null} error
 */

export {};
